### PR TITLE
Refactor transaction validation in block normalization - Closes #2905

### LINKED ISSUE #2905: Refactor validating transactions using elements transactions library
### Expected behavior

In current implementation, transactions are validated using `objectNormalize` in `logic/transaction` when receiving new block. 

### Actual behavior

As part of the implementation process of `elements transactions` library into core, we will replace legacy validation functionality in core.

### Which version(s) does this affect? (Environment, OS, etc...)

1.6

### PR DESCRIPTION
**Pull request by mitsujutsu**
### What was the problem?

In current implementation, transactions are validated using `objectNormalize` in `logic/transaction` when receiving new block.

### How did I fix it?

Refactored block verify `normalizeBlock` and loader `loadTransactions` to use ProcessTransactions module.

### How to test it?

Run tests for `verify.js`

### Review checklist

* The PR resolves #2905 
* All new code is covered with unit tests
* All new code was formatted with Prettier
* Linting passes
* Tests pass
* Commit messages follow the [commit guidelines](CONTRIBUTING.md#git-commit-messages)
* Documentation has been added/updated

### CODE REVIEW

File: framework/test/common/utils/random.js [hunk 1]
```diff
@@ -86,7 +86,7 @@ random.application = function() {
 		name: random.applicationName(),
 		description: 'Blockchain based home monitoring tool',
 		tags: 'monitoring temperature power sidechain',
-		type: random.number(0, 2),
+		type: 0,
```
Comment (shuse2) [inline framework/test/common/utils/random.js]: im not sure why it was random, but should it not be random?

File: framework/src/modules/chain/api/controllers/node.js [hunk 2]
```diff
@@ -57,8 +57,12 @@ function NodeController(scope) {
 	};
 	blockReward = new BlockReward();
 	getNetworkHeight = promisify(library.modules.peers.networkHeight);
-	getTransactionsCount = promisify(library.modules.transactions.shared.getTransactionsCount);
-	updateForgingStatus = promisify(library.modules.delegates.updateForgingStatus);
```
Comment (MaciejBaj) [inline framework/src/modules/chain/api/controllers/node.js]: Again, something is wrong with the formatting on your branch or maybe the whole feature/improve_transactions_processing_efficiency. Please solve this @mitsujutsu.

Comment (shuse2) [reply framework/src/modules/chain/api/controllers/node.js]: It seems like it's happening all the time 😕

File: framework/src/modules/chain/logic/block.js [hunk 3]
```diff
@@ -242,16 +242,6 @@ class Block {
 				.join(', ')}`;
 		}
 
-		try {
-			for (let i = 0; i < block.transactions.length; i++) {
-				block.transactions[i] = this.scope.transaction.objectNormalize(
```
Comment (MaciejBaj) [inline framework/src/modules/chain/logic/block.js]: Why there is no need for normalization here?

File: framework/test/common/utils/random.js (hunk 1 above)
Comment (mitsujutsu) [reply framework/test/common/utils/random.js]: `dapp.type` required to be `0` in core and elements logic, therefore tests using this random function will occasionally fail if left `random(0,2)`

File: framework/src/modules/chain/modules/loader.js [hunk 4]
```diff
@@ -286,36 +287,35 @@ __private.loadTransactions = function(cb) {
...
+					);
+					const invalidTransactionResponse = transactionResponses.find(
+						transactionResponse =>
+							transactionResponse.status !== TransactionStatus.OK
+					);
+					if (invalidTransactionResponse) {
+						throw invalidTransactionResponse.errors;
+					}
+				} catch (e) {
+					const error = Array.isArray(e) && e.length > 0 ? e[0] : e;
```
[the remark this answers is not in the document]
Comment (mitsujutsu) [reply framework/src/modules/chain/modules/loader.js]: it is also due to `initTransaction` in `validateTransactions`(basically the constructor for each elements transaction class) also throwing an error .

File: framework/src/modules/chain/modules/loader.js [hunk 5]
```diff
@@ -286,36 +287,35 @@ __private.loadTransactions = function(cb) {
 				});
 			},
 			function(peer, transactions, waterCb) {
-				async.eachSeries(
-					transactions,
-					(transaction, eachSeriesCb) => {
-						const id = transaction ? transactions.id : 'null';
+				try {
+					const transactionResponses = library.modules.processTransactions.validateTransactions(
```
[the remark this answers is not in the document]
Comment (shuse2) [reply framework/src/modules/chain/modules/loader.js]: transactionResponses is correct here i think, actual type is TransactionResponse[];

File: framework/test/common/utils/random.js (hunk 1 above)
Comment (shuse2) [reply framework/test/common/utils/random.js]: this part needs to be put it back after the fix in elements.

File: framework/src/modules/chain/logic/block.js [hunk 6]
```diff
@@ -232,24 +234,27 @@ class Block {
...
+				transactionResponse =>
+					transactionResponse.status !== TransactionStatus.OK
+			);
+			if (invalidTransactionResponse) {
+				throw invalidTransactionResponse.errors;
 			}
-		} catch (e) {
-			throw e;
+		} catch (errors) {
+			const error = Array.isArray(errors) && errors.length > 0 ? errors[0] : errors;
```
Comment (pablitovicente) [inline framework/src/modules/chain/logic/block.js]: Why only first error here if it's an array?

File: framework/src/modules/chain/modules/loader.js [hunk 7]
```diff
@@ -286,36 +287,36 @@ __private.loadTransactions = function(cb) {
...
+					const invalidTransactionResponse = transactionResponses.find(
+						transactionResponse =>
+							transactionResponse.status !== TransactionStatus.OK
+					);
+					if (invalidTransactionResponse) {
+						throw invalidTransactionResponse.errors;
+					}
+				} catch (errors) {
+					const error =
+						Array.isArray(errors) && errors.length > 0 ? errors[0] : errors;
```
Comment (pablitovicente) [inline framework/src/modules/chain/modules/loader.js]: Same here why we return the first one in the case of errors being an array?

File: framework/test/integration/process_transactions.js [hunk 8]
```diff
@@ -81,42 +82,72 @@ describe('blocks/verifyTransactions', async () => {
...
+			const transactionResponses = await verifyTransactions(
+				nonVerifiableTransactions
+			);
+			transactionResponses.forEach(transactionResponse => {
+				expect(transactionResponse.status).to.equal(transactionStatus.FAIL);
+			});
+		});
+
+		/* eslint-disable mocha/no-skipped-tests */
+		it.skip('should return transactionResponse with status PENDING for transactions waiting multi-signatures', async () => {
```
Comment (pablitovicente) [inline framework/test/integration/process_transactions.js]: why is this one skipped?

File: framework/src/modules/chain/logic/block.js (hunk 6 above)
Comment (shuse2) [reply framework/src/modules/chain/logic/block.js]: I think originally it was only returning one error

File: framework/src/modules/chain/modules/blocks/verify.js [hunk 9]
```diff
@@ -618,8 +618,8 @@ __private.addBlockProperties = function(block, broadcast, cb) {
 __private.normalizeBlock = function(block, cb) {
 	try {
 		block = library.logic.block.objectNormalize(block);
-	} catch (err) {
-		return setImmediate(cb, err);
+	} catch (errors) {
```
Comment (SargeKhan) [inline framework/src/modules/chain/modules/blocks/verify.js]: We are only throwing a single error from `block.objectNormalize` function.

File: framework/test/mocha/unit/modules/chain/modules/transport.js [hunk 10]
```diff
@@ -744,58 +739,6 @@ describe('transport', () => {
...
-					).to.be.true;
-				});
-
-				it('should call callback with error = "Invalid transaction body"', async () =>
-					expect(error).to.equal(
-						`Invalid transaction body - ${objectNormalizeError}`
-					));
-			});
-
 			describe('when nonce is undefined', () => {
```
Comment (SargeKhan) [inline framework/test/mocha/unit/modules/chain/modules/transport.js]: We should at least have one test for checking that `validateTransactions` is called for transactions in the block.

File: framework/src/modules/chain/modules/process_transactions.js [hunk 11]
```diff
@@ -29,6 +29,18 @@ class ProcessTransactions {
...
 
+	// eslint-disable-next-line class-methods-use-this
+	validateTransactions(transactions) {
+		return transactions
+			.map(transaction => ({
+				...transaction,
+				fee: transaction.fee.toString(),
+				amount: transaction.amount.toString(),
+			}))
+			.map(initTransaction)
```
[the remark this answers is not in the document]
Comment (pablitovicente) [reply framework/src/modules/chain/modules/process_transactions.js]: Or maybe change the name of the method only to avoid refactoring at this stage

**Pull request merged**